dask/layers.py

```diff
@@ -103,6 +105,165 @@ def __init__(
         )
 
 
+class SlicingLayer(Layer):
+    """Array slicing HighLevelGraph Layer"""
+
+    def __init__(
+        self,
+        out_name,
+        in_name,
+        blockdims,
+        new_blockdims,
+        index,
+        shape,
+        parts_out=None,
+    ):
+        super().__init__()
+        self.out_name = out_name
+        self.in_name = in_name
+        self.blockdims = blockdims
+        self.new_blockdims = new_blockdims
+        self.index = index
+        self.shape = shape
+        self.parts_out = parts_out
+
+    def __repr__(self):
+        return "SlicingLayer<name='{}'".format(self.out_name)
+
+    @property
+    def _dict(self):
+        """Materialize full dict representation"""
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict
+        else:
+            dsk = self._construct_graph()
+            self._cached_dict = dsk
+        return self._cached_dict
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
+
+    def is_materialized(self):
+        return hasattr(self, "_cached_dict")
+
+    def get_output_keys(self):
+        return self.keys()  # FIXME! this implementation materializes the graph
+
+    def _get_block_slices(self, shape=None, blockdims=None, index=None):
+        """Get a list (for each dimension) of dicts{blocknum: slice()}"""
+        try:
+            block_slices = self._block_slices
+        except AttributeError:
+            block_slices = list(map(_slice_1d, shape, blockdims, index))
+            self._block_slices = block_slices
+        return block_slices
+
+    def _keys_to_output_parts(self, keys):
+        """Simple utility to convert keys to array chunk indices."""
+        parts = set()
+        for key in keys:
+            try:
+                _name, *_part = key
+            except ValueError:
+                continue
+            if _name != self.out_name:
+                continue
+            parts.add(tuple(_part))
+        return parts
+
+    def _input_parts(self):
+        """Simple utility to get input chunk indices."""
+        parts_in = set()
+        block_slices = self._get_block_slices(self.shape, self.blockdims, self.index)
+        _part = []
+        for i in block_slices:
+            _part += sorted(i.keys())
+        parts_in.add(tuple(_part))
+        return parts_in
```

Comment:
This is very nice.

Reply:
Sadly, no. This is the bit that's currently broken & I can't see a way to fix it without using the itertools product. (I know I said I fixed it, I was wrong. It only happened to work well for the one specific example I used as a test case)

I might have explained myself poorly today, perhaps this summary helps:
* You can see how this implementation breaks things, but choosing any slicing example that will return values *from more than one input chunk*. Eg;
```python
import numpy as np
import dask.array as da

x = da.from_array(np.arange(16).reshape((4,4)), chunks=(2,2))
y = x[1:, 1:]
y.compute()  # fails with the implementation
```
* The itertools product function is needed so we list all of the necessary input chunks (all the combinations from our 1d chunk results).
* This means there's not a lot of logic left that only happens in the `_construct_graph` function. That's why I was concerned about how useful this is (but I hadn't been considering your point about the benefits for nested slicing)

Reply:
I've just pushed a commit that replaces
```python
    def _input_parts(self):
        """Simple utility to get input chunk indices."""
        parts_in = set()
        block_slices = self._get_block_slices(self.shape, self.blockdims, self.index)
        _part = []
        for i in block_slices:
            _part += sorted(i.keys())
        parts_in.add(tuple(_part))
        return parts_in
```

with this instead

```python
    def _input_parts(self):
        """Simple utility to get input chunk indices."""
        block_slices = self._get_block_slices(self.shape, self.blockdims, self.index)
        parts_in = set(product(*[sorted(i.keys()) for i in block_slices]))
        return parts_in
```

Now I can make slices over multiple chunks again (yay, less broken), but it's not a great sign for the performance improvements we were hoping to get out of this. I haven't benchmarked anything yet though.